File: src/main/java/com/team10/backend/domain/image/service/ImageUploadService.java
package com.team10.backend.domain.image.service;

import com.team10.backend.domain.image.dto.ImageUploadResponse;
import com.team10.backend.domain.image.dto.PresignedUrlRequest;
import com.team10.backend.domain.image.dto.PresignedUrlResponse;
import com.team10.backend.global.exception.BusinessException;
import com.team10.backend.global.exception.ErrorCode;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import org.springframework.web.multipart.MultipartFile;
import software.amazon.awssdk.core.exception.SdkException;
import software.amazon.awssdk.core.sync.RequestBody;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import software.amazon.awssdk.services.s3.presigner.S3Presigner;
import software.amazon.awssdk.services.s3.presigner.model.PresignedPutObjectRequest;
import software.amazon.awssdk.services.s3.presigner.model.PutObjectPresignRequest;

import java.io.IOException;
import java.net.URI;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.Set;
import java.util.UUID;

//import lombok.extern.slf4j.Slf4j;

@Service
//@Slf4j
public class ImageUploadService {

    private static final Duration PRESIGNED_URL_DURATION = Duration.ofMinutes(5);

    private static final Set<String> ALLOWED_CONTENT_TYPES = Set.of(
            "image/jpeg",
            "image/png",
            "image/webp",
            "image/gif"
    );

    private final S3Client s3Client;
    private final S3Presigner s3Presigner;

    @Value("${cloud.aws.s3.bucket}")
    private String bucket;

    @Value("${cloud.aws.region}")
    private String region;

    // 이미지 파일을 검증한 뒤 S3에 업로드하고, DB에 저장할 수 있는 imageUrl을 반환한다.
    public ImageUploadResponse upload(MultipartFile file, String directory) {
        validate(file);

        String key = createObjectKey(file, directory);
        String bucketName = bucket.trim();

        // S3에 저장할 버킷, 객체 key, 파일 타입/크기를 지정한다.
        PutObjectRequest request = PutObjectRequest.builder()
                .bucket(bucketName)
                .key(key)
                .contentType(file.getContentType())
                .contentLength(file.getSize())
                .build();

        try {
            s3Client.putObject(request, RequestBody.fromInputStream(file.getInputStream(), file.getSize()));
        } catch (IOException | SdkException e) {
//            log.error("S3 image upload failed. bucket={}, key={}, reason={}", bucketName, key, e.getMessage(), e);
            throw new BusinessException(ErrorCode.FILE_UPLOAD_FAILED);
        }

        return new ImageUploadResponse(createImageUrl(key));
    }

    // 다중 이미지 파일을 S3에 업로드하고, DB에 저장할 수 있는 리스트를 반환
    public List<ImageUploadResponse> uploadMultiple(List<MultipartFile> files, String directory) {
        validateFiles(files);

        return files.stream()
                .map(file -> upload(file, directory))
                .toList();
    }

    public PresignedUrlResponse createPresignedUrl(PresignedUrlRequest request) {
        validatePresignedUrlRequest(request);

        String key = createObjectKey(request.fileName(), request.directory());
        String bucketName = bucket.trim();

        PutObjectRequest putObjectRequest = PutObjectRequest.builder()
                .bucket(bucketName)
                .key(key)
                .contentType(request.contentType())
                .build();

        PutObjectPresignRequest presignRequest = PutObjectPresignRequest.builder()
                .signatureDuration(PRESIGNED_URL_DURATION)
                .putObjectRequest(putObjectRequest)
                .build();

        PresignedPutObjectRequest presignedRequest = s3Presigner.presignPutObject(presignRequest);

        return new PresignedUrlResponse(
                presignedRequest.url().toString(),
                createImageUrl(key)
        );
    }

    // imageUrl에서 S3 object key를 추출해 실제 S3 파일을 삭제한다.
    public void delete(String imageUrl) {
        if (!StringUtils.hasText(imageUrl)) {
            return;
        }

        String key = extractObjectKey(imageUrl);
        String bucketName = bucket.trim();

        DeleteObjectRequest request = DeleteObjectRequest.builder()
                .bucket(bucketName)
                .key(key)
                .build();
        try {
            s3Client.deleteObject(request);
        } catch (SdkException e) {
//            log.error("S3 image delete failed. bucket={}, key={}, reason={}", bucketName, key, e.getMessage(), e);
            throw new BusinessException(ErrorCode.FILE_DELETE_FAILED);
        }
    }

    public void deleteMultiple(List<String> imageUrls) {
        validateImageUrls(imageUrls);
        imageUrls.forEach(this::deleteIfManaged);
    }

    // 피드/상품 수정 시 기존 이미지가 우리 S3 URL인 경우에만 삭제한다.
    public void deleteIfManaged(String imageUrl) {
        if (!isManagedImageUrl(imageUrl)) {
            return;
        }

        delete(imageUrl);
    }

    // 버킷 설정, 빈 파일 여부, 허용된 이미지 MIME 타입인지 확인한다.
    private void validate(MultipartFile file) {
        if (!StringUtils.hasText(bucket)) {
            throw new BusinessException(ErrorCode.FILE_UPLOAD_FAILED, "S3 버킷 설정이 없습니다.");
        }
        if (file == null || file.isEmpty()) {
            throw new BusinessException(ErrorCode.INVALID_IMAGE_FILE, "업로드할 이미지 파일이 없습니다.");
        }
        if (!ALLOWED_CONTENT_TYPES.contains(file.getContentType())) {
            throw new BusinessException(ErrorCode.INVALID_IMAGE_FILE);
        }
    }

    private void validatePresignedUrlRequest(PresignedUrlRequest request) {
        if (!StringUtils.hasText(bucket)) {
            throw new BusinessException(ErrorCode.FILE_UPLOAD_FAILED, "S3 버킷 설정이 없습니다.");
        }
        if (request == null || !StringUtils.hasText(request.fileName())) {
            throw new BusinessException(ErrorCode.INVALID_IMAGE_FILE, "파일명은 필수입니다.");
        }
        if (!ALLOWED_CONTENT_TYPES.contains(request.contentType())) {
            throw new BusinessException(ErrorCode.INVALID_IMAGE_FILE);
        }
    }

    private void validateFiles(List<MultipartFile> files) {
        if (files == null || files.isEmpty()) {
            throw new BusinessException(ErrorCode.INVALID_IMAGE_FILE, "업로드할 이미지 파일이 없습니다.");
        }

        if (files.size() > 10) {
            throw new BusinessException(ErrorCode.INVALID_IMAGE_FILE, "파일은 최대 10개 까지 업로드할 수 있습니다.");
        }
    }

    private void validateImageUrls(List<String> imageUrls) {
        if (imageUrls == null || imageUrls.isEmpty()) {
            throw new BusinessException(ErrorCode.INVALID_INPUT, "삭제할 이미지가 없습니다.");
        }
    }


    // S3 object key를 만든다. 예: feeds/{uuid}.jpg
    private String createObjectKey(MultipartFile file, String directory) {
        String safeDirectory = sanitizeDirectory(directory);
        String extension = getExtension(file.getOriginalFilename());
        return safeDirectory + "/" + UUID.randomUUID() + extension;
    }

    private String createObjectKey(String fileName, String directory) {
        String safeDirectory = sanitizeDirectory(directory);
        String extension = getExtension(fileName);
        return safeDirectory + "/" + UUID.randomUUID() + extension;
    }

    // directory 값에서 위험하거나 불필요한 문자를 제거해 S3 prefix로 안전하게 사용한다.
    private String sanitizeDirectory(String directory) {
        if (!StringUtils.hasText(directory)) {
            return "images";
        }
        String sanitized = directory.replaceAll("[^a-zA-Z0-9/_-]", "");
        sanitized = sanitized.replaceAll("^/+", "").replaceAll("/+$", "");
        return StringUtils.hasText(sanitized) ? sanitized : "images";
    }

    // 원본 파일명에서 확장자만 추출해 UUID 파일명 뒤에 붙인다.
    private String getExtension(String filename) {
        if (!StringUtils.hasText(filename)) {
            return "";
        }
        String cleanFilename = StringUtils.cleanPath(filename);
        int extensionIndex = cleanFilename.lastIndexOf(".");
        return extensionIndex >= 0 ? cleanFilename.substring(extensionIndex).toLowerCase() : "";
    }

    // 우리 S3 버킷 URL에서만 object key를 뽑아낸다. 예: https://bucket.s3.region.amazonaws.com/feeds/a.jpg -> feeds/a.jpg
    private String extractObjectKey(String imageUrl) {
        try {
            URI uri = URI.create(imageUrl.trim());
            if (!isManagedImageUrl(uri)) {
                throw new BusinessException(ErrorCode.INVALID_INPUT, "우리 S3 버킷 이미지 URL만 삭제할 수 있습니다.");
            }

            String path = uri.getPath();
            if (!StringUtils.hasText(path) || "/".equals(path)) {
                throw new BusinessException(ErrorCode.INVALID_INPUT, "삭제할 이미지 경로가 없습니다.");
            }

            return URLDecoder.decode(path.substring(1), StandardCharsets.UTF_8);
        } catch (IllegalArgumentException e) {
            throw new BusinessException(ErrorCode.INVALID_INPUT, "이미지 URL 형식이 올바르지 않습니다.");
        }
    }

    private boolean isManagedImageUrl(String imageUrl) {
        if (!StringUtils.hasText(imageUrl)) {
            return false;
        }

        try {
            return isManagedImageUrl(URI.create(imageUrl.trim()));
        } catch (IllegalArgumentException e) {
            return false;
        }
    }

    private boolean isManagedImageUrl(URI uri) {
        String expectedHost = bucket.trim() + ".s3." + region.trim() + ".amazonaws.com";
        return expectedHost.equals(uri.getHost());
    }

    // S3에 저장된 object key를 기반으로 클라이언트가 사용할 imageUrl을 만든다.
    private String createImageUrl(String key) {
        return "https://" + bucket.trim() + ".s3." + region.trim() + ".amazonaws.com/" + key;
    }

    public ImageUploadService(S3Client s3Client, S3Presigner s3Presigner) {
        this.s3Client = s3Client;
        this.s3Presigner = s3Presigner;
    }
}

File: src/main/kotlin/com/team10/backend/domain/image/service/ImageUploadService.kt
package com.team10.backend.domain.image.service

import com.team10.backend.domain.image.dto.PresignedUrlRequest
import com.team10.backend.domain.image.dto.PresignedUrlResponse
import com.team10.backend.global.exception.BusinessException
import com.team10.backend.global.exception.ErrorCode
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.util.StringUtils
import software.amazon.awssdk.core.exception.SdkException
import software.amazon.awssdk.services.s3.S3Client
import software.amazon.awssdk.services.s3.model.DeleteObjectRequest
import software.amazon.awssdk.services.s3.model.PutObjectRequest
import software.amazon.awssdk.services.s3.presigner.S3Presigner
import software.amazon.awssdk.services.s3.presigner.model.PutObjectPresignRequest
import java.net.URI
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.time.Duration
import java.util.*

//import lombok.extern.slf4j.Slf4j;
@Service //@Slf4j
class ImageUploadService(private val s3Client: S3Client, private val s3Presigner: S3Presigner) {
    @Value("\${cloud.aws.s3.bucket}")
    private lateinit var bucket: String

    @Value("\${cloud.aws.region}")
    private lateinit var region: String

    fun createPresignedUrl(request: PresignedUrlRequest): PresignedUrlResponse {
        validatePresignedUrlRequest(request)

        val key = createObjectKey(request.fileName, request.directory)
        val bucketName = bucket.trim()

        val putObjectRequest = PutObjectRequest.builder()
            .bucket(bucketName)
            .key(key)
            .contentType(request.contentType)
            .build()

        val presignRequest = PutObjectPresignRequest.builder()
            .signatureDuration(PRESIGNED_URL_DURATION)
            .putObjectRequest(putObjectRequest)
            .build()

        val presignedRequest = s3Presigner.presignPutObject(presignRequest)

        return PresignedUrlResponse(
            presignedRequest.url().toString(),
            createImageUrl(key)
        )
    }

    // imageUrl에서 S3 object key를 추출해 실제 S3 파일을 삭제한다.
    fun delete(imageUrl: String?) {
        if (!StringUtils.hasText(imageUrl)) {
            return
        }

        val key = extractObjectKey(imageUrl!!)
        val bucketName = bucket.trim()

        val request = DeleteObjectRequest.builder()
            .bucket(bucketName)
            .key(key)
            .build()
        try {
            s3Client.deleteObject(request)
        } catch (e: SdkException) {
//            log.error("S3 image delete failed. bucket={}, key={}, reason={}", bucketName, key, e.getMessage(), e);
            throw BusinessException(ErrorCode.FILE_DELETE_FAILED)
        }
    }

    // 피드/상품 수정 시 기존 이미지가 우리 S3 URL인 경우에만 삭제한다.
    fun deleteIfManaged(imageUrl: String?) {
        if (!isManagedImageUrl(imageUrl)) {
            return
        }

        delete(imageUrl)
    }


    private fun validatePresignedUrlRequest(request: PresignedUrlRequest) {
        if (!StringUtils.hasText(bucket)) {
            throw BusinessException(ErrorCode.FILE_UPLOAD_FAILED, "S3 버킷 설정이 없습니다.")
        }
        if (!StringUtils.hasText(request.fileName)) {
            throw BusinessException(ErrorCode.INVALID_IMAGE_FILE, "파일명은 필수입니다.")
        }
        if (!ALLOWED_CONTENT_TYPES.contains(request.contentType)) {
            throw BusinessException(ErrorCode.INVALID_IMAGE_FILE)
        }
    }

    private fun createObjectKey(fileName: String, directory: String?): String {
        val safeDirectory = sanitizeDirectory(directory)
        val extension = getExtension(fileName)
        return safeDirectory + "/" + UUID.randomUUID() + extension
    }

    // directory 값에서 위험하거나 불필요한 문자를 제거해 S3 prefix로 안전하게 사용한다.
    private fun sanitizeDirectory(directory: String?): String {
        if (!StringUtils.hasText(directory)) {
            return "images"
        }
        var sanitized = directory!!.replace("[^a-zA-Z0-9/_-]".toRegex(), "")
        sanitized = sanitized.replace("^/+".toRegex(), "").replace("/+$".toRegex(), "")
        return if (StringUtils.hasText(sanitized)) sanitized else "images"
    }

    // 원본 파일명에서 확장자만 추출해 UUID 파일명 뒤에 붙인다.
    private fun getExtension(filename: String): String {
        if (!StringUtils.hasText(filename)) {
            return ""
        }
        val cleanFilename = StringUtils.cleanPath(filename)
        val extensionIndex = cleanFilename.lastIndexOf(".")
        return if (extensionIndex >= 0) cleanFilename.substring(extensionIndex).lowercase(Locale.getDefault()) else ""
    }

    // 우리 S3 버킷 URL에서만 object key를 뽑아낸다. 예: https://bucket.s3.region.amazonaws.com/feeds/a.jpg -> feeds/a.jpg
    private fun extractObjectKey(imageUrl: String): String {
        try {
            val uri = URI.create(imageUrl.trim { it <= ' ' })
            if (!isManagedImageUrl(uri)) {
                throw BusinessException(ErrorCode.INVALID_INPUT, "우리 S3 버킷 이미지 URL만 삭제할 수 있습니다.")
            }

            val path = uri.getPath()
            if (!StringUtils.hasText(path) || "/" == path) {
                throw BusinessException(ErrorCode.INVALID_INPUT, "삭제할 이미지 경로가 없습니다.")
            }

            return URLDecoder.decode(path.substring(1), StandardCharsets.UTF_8)
        } catch (e: IllegalArgumentException) {
            throw BusinessException(ErrorCode.INVALID_INPUT, "이미지 URL 형식이 올바르지 않습니다.")
        }
    }

    private fun isManagedImageUrl(imageUrl: String?): Boolean {
        if (!StringUtils.hasText(imageUrl)) {
            return false
        }

        try {
            return isManagedImageUrl(URI.create(imageUrl!!.trim()))
        } catch (e: IllegalArgumentException) {
            return false
        }
    }

    private fun isManagedImageUrl(uri: URI): Boolean {
        val expectedHost = bucket.trim() + ".s3." + region.trim() + ".amazonaws.com"
        return expectedHost == uri.getHost()
    }

    // S3에 저장된 object key를 기반으로 클라이언트가 사용할 imageUrl을 만든다.
    private fun createImageUrl(key: String): String {
        return "https://" + bucket.trim() + ".s3." + region.trim() + ".amazonaws.com/" + key
    }

    companion object {
        private val PRESIGNED_URL_DURATION: Duration = Duration.ofMinutes(5)

        private val ALLOWED_CONTENT_TYPES = setOf(
            "image/jpeg",
            "image/png",
            "image/webp",
            "image/gif"
        )
    }
}
